### ISSUE TITLE
[BUG] SqlAlchemyStore search_runs treats a NaN metric as 0 in numeric filters, so `metrics.loss < 0.1` returns diverged runs

### ISSUE THREAD

**Issue by 2sumtech** [mlflow/mlflow#25640]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

**Issues Policy acknowledgement**

- [x] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

**Where did you encounter this bug?**

Local machine

**Willingness to contribute**

Yes. I can contribute a fix for this bug independently.

**MLflow version**

`mlflow` master (`3.16.1.dev0`, commit `0044946`)

**System information**

- OS Platform and Distribution: macOS 15 (Darwin 25.6.0)
- Python version: 3.10

**Describe the problem**

With the SQLAlchemy tracking store, a run whose metric is `NaN` matches numeric filters as if the value were `0`. `search_runs(filter_string="metrics.loss < 0.1")` returns a run whose `loss` is `NaN`, and so does `metrics.loss = 0`.

`SqlAlchemyStore` stores a NaN metric as `value=0, is_nan=True` (see `sanitize_metric_value` in `mlflow/store/tracking/sqlalchemy_store.py`). The metric filter in `_get_sqlalchemy_filter_clauses` then compares `SqlLatestMetric.value` against the filter value with no `is_nan` guard, so the placeholder `0` takes part in the comparison. The order-by path in the same file already treats NaN as its own category (`_get_orderby_clauses` emits a `CASE` on `is_nan` so NaN sorts last, covered by tests), so filtering is the one place where the two-column representation leaks.

This matters because filtering on metrics is how people pick the best runs. `metrics.loss < 0.1` ranks diverged runs as the best ones, silently, on the recommended database backend. The file store compares the real `nan`, so it returns nothing for these filters, which matches IEEE 754 (every comparison against NaN is false) and matches what the same run reports through `run.data.metrics`.

**Code to reproduce issue**

```python
import tempfile, mlflow
from mlflow import MlflowClient

tmp = tempfile.mkdtemp()
mlflow.set_tracking_uri(f"sqlite:///{tmp}/mlflow.db")
c = MlflowClient()
exp = c.create_experiment("nan-demo")
for name, loss in [("good", 0.5), ("diverged", float("nan")), ("bad", 9.0)]:
    r = c.create_run(exp, run_name=name)
    c.log_metric(r.info.run_id, "loss", loss)
for f in ["metrics.loss < 0.1", "metrics.loss = 0", "metrics.loss < 1"]:
    print(f, "->", sorted(r.info.run_name for r in c.search_runs([exp], f)))
print(c.search_runs([exp], "attributes.run_name = 'diverged'")[0].data.metrics)
```

**Stack trace / observed output**

```
sqlite  metrics.loss < 0.1     -> ['diverged']
sqlite  metrics.loss = 0       -> ['diverged']
sqlite  metrics.loss < 1       -> ['diverged', 'good']
reported loss of diverged: {'loss': nan}
```

Same script against a file store (`MLFLOW_ALLOW_FILE_STORE=true`):

```
file    metrics.loss < 0.1     -> []
file    metrics.loss = 0       -> []
file    metrics.loss < 1       -> ['good']
reported loss of diverged: {'loss': nan}
```

**Other info / logs**

Expected: `metrics.loss < 0.1` and `metrics.loss = 0` return no runs, `metrics.loss < 1` returns only `good`, on every backend.

Proposed fix: in the metric filter clause, add `SqlLatestMetric.is_nan == false()` alongside the value comparison for numeric comparators, so a NaN placeholder can never satisfy `<`, `<=`, `>`, `>=`, `=`. `!=` is the one comparator where NaN should match, which mirrors IEEE semantics. Happy to send the PR with a regression test once this is triaged.

Related but different: #17270 is about the UI showing the sanitized `0` for NaN metrics on the Models page. This issue is about the search filter on the server side.

**What component(s) does this bug affect?**

- [x] `area/tracking`

**Comment by ayushh0110** [comment]
Hey, I'd be interested in picking this up!

I traced through the code and the bug is pretty clear — `sanitize_metric_value` stores NaN as `value=0, is_nan=True`, and the order-by path already handles this correctly (there's a `CASE` on `is_nan` so NaN sorts last), but the filter path in `_get_sqlalchemy_filter_clauses` just compares `entity.value` directly without checking `is_nan`. So that placeholder `0` ends up matching filters it shouldn't.

The fix should be straightforward: when building the metric filter subquery, tack on `SqlLatestMetric.is_nan == false()` for all comparators except `!=`. That way NaN rows can't satisfy `<`, `<=`, `>`, `>=`, or `=` — which lines up with IEEE 754 and how the file store already behaves. For `!=` we'd skip the guard since `NaN != x` is true for any x.

I'll include a regression test covering the cases from the repro (`metrics.loss < 0.1` returning empty, `metrics.loss = 0` returning empty, `metrics.loss < 1` returning only the good run, etc.) plus a `!=` case to make sure NaN still matches there.

Could a maintainer triage this and apply the `ready` label so I can open a PR? Thanks!

**Comment by 2sumtech** [comment]
Thanks for the interest. As noted in the report I am contributing this one myself and the fix plus regression test are already done on my fork at https://github.com/2sumtech/mlflow/tree/fix/nan-metric-filter. It guards the numeric comparators with is_nan == false and lets != match NaN rows, with a test covering the cases from the repro and the != case. I will open the PR as soon as a maintainer applies the ready label.

**Comment by kriscon-db** [comment]
Thanks for the detailed report, and for offering to take this on. We reproduced it on master against a SQLite backend: a run with `loss=NaN` (stored as `value=0, is_nan=True`) matches `metrics.loss < 0.1` and `metrics.loss = 0`, exactly as you described.

Your root-cause analysis is correct. `_get_sqlalchemy_filter_clauses` builds the metric subquery filter with no `is_nan` guard, so the placeholder `0` participates in the numeric comparison, while `_get_orderby_clauses` already treats NaN as its own category via a `CASE` on `is_nan`.

Marking this `ready`, it's yours if you'd like to pick it up. One thing worth handling in the fix: to match the file-store / IEEE 754 behavior, a NaN metric should not match `<`, `<=`, `>`, `>=`, or `=`, but `!=` should still return it (`NaN != x` is true), so the guard needs to be comparator-aware rather than a blanket exclusion. A parametrized test covering all comparators would be great.

-- Claude-drafted, reviewed before posting.
